src/openmc_heat_driver.cpp

```diff
@@ -100,36 +125,143 @@ void OpenmcHeatDriver::init_mappings()
+  // the surrogate T/H model is sufficiently fine in the pellet region to
+  // capture all fissionable cells in OpenMC.
+  Ensures(openmc_driver_->n_fissionable_cells_ == n_fissionable_mapped);
+
+  // set the number of solid cells before defining mappings for fluid cells
   if (openmc_driver_->active()) {
-    n_materials_ = openmc_driver_->cells_.size();
+    n_solid_cells_ = openmc_driver_->cells_.size();
+  }
+
+  // Establish mappings between fluid regions and OpenMC cells; it is assumed that there
+  // are no azimuthal divisions in the fluid phase in the OpenMC model so that we
+  // can take a point on a 45 degree ray from the pin center. TODO: add a check to make
+  // sure that the T/H model is finer than the OpenMC model.
+  for (int i = 0; i < heat_driver_->n_pins_; ++i) {
+    double x_center = heat_driver_->pin_centers_(i, 0);
+    double y_center = heat_driver_->pin_centers_(i, 1);
+
+    for (int j = 0; j < heat_driver_->n_axial_; ++j) {
+      double zavg = 0.5 * (z(j) + z(j + 1));
+      double l = heat_driver_->pin_pitch() / std::sqrt(2.0);
+      double d = (l - heat_driver_->clad_outer_radius_) / 2.0;
+      double x = x_center + (heat_driver_->clad_outer_radius_ + d) * std::sqrt(2.0) / 2.0;
+      double y = y_center + (heat_driver_->clad_outer_radius_ + d) * std::sqrt(2.0) / 2.0;
+
+      // Determine cell instance corresponding to given fluid location
+      Position r{x, y, zavg};
+      CellInstance c{r};
+      if (tracked.find(c) == tracked.end()) {
+        openmc_driver_->cells_.push_back(c);
+        tracked[c] = openmc_driver_->cells_.size() - 1;
+
+        Ensures(!c.is_fissionable());
+      }
+
+      // Map OpenMC material to fluid element and vice versa
+      int32_t array_index = tracked[c];
+      cell_inst_to_elem_[array_index].push_back(elem_index);
+      elem_to_cell_inst_[elem_index].push_back(array_index);
+
+      ++elem_index;
+    }
+  }
+
+  if (openmc_driver_->active()) {
+    n_fluid_cells_ = openmc_driver_->cells_.size() - n_solid_cells_;
   }
 }
 
 void OpenmcHeatDriver::init_tallies()
 {
   if (openmc_driver_->active()) {
-    // Build vector of material indices
+    // Build vector of material indices to construct tallies; tallies are only
+    // used in the solid regions
     std::vector<int32_t> mats;
-    for (const auto& c : openmc_driver_->cells_) {
-      mats.push_back(c.material_index_);
+    for (int c = 0; c < n_solid_cells_; ++c) {
+      const auto& cell = openmc_driver_->cells_[c];
+      mats.push_back(cell.material_index_);
     }
+
     openmc_driver_->create_tallies(mats);
   }
 }
 
+void OpenmcHeatDriver::init_densities()
+{
+  if (this->has_global_coupling_data()) {
+    std::size_t n_solid =
+      heat_driver_->n_pins_ * heat_driver_->n_axial_ * heat_driver_->n_rings();
+    std::size_t n_fluid = heat_driver_->n_pins_ * heat_driver_->n_axial_;
+    densities_.resize({n_solid + n_fluid});
+    densities_prev_.resize({n_solid + n_fluid});
```

Comment:
Why not just make this of size `{n_fluid}`?

Reply:
I left it as size equal to the total T/H elements for similarity with the Nek mappings. I think what I'd prefer is to change `HeatFluidsDriver` so that the interpretation of density is determined by the derived class (i.e. for Nek, it'd be solid and fluid densities mixed together, while for the surrogate it might only be fluid density). Because that would touch code aspects unrelated to subchannel things I thought that could go as a separate _PR_ :)

Reply:
Ah, makes sense. Thanks, look forward to your next _PR_ :laughing: